controllers/wso2is_controller.go

```diff
@@ -0,0 +1,592 @@
+		return ctrl.Result{}, err
+	}
+	podNames := getPodNames(podList.Items)
+
+	// Update status.Nodes if needed
+	if !reflect.DeepEqual(podNames, instance.Status.Nodes) {
+		instance.Status.Nodes = podNames
+		err := r.Status().Update(ctx, &instance)
+		if err != nil {
+			log.Error(err, "Failed to update WSO2IS status")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// labelsForWso2IS returns the labels for selecting the resources
+// belonging to the given WSO2IS CR name.
+func labelsForWso2IS(depname string) map[string]string {
+	return map[string]string{
+		"deployment": depname,
+		"app":        depname,
+		"monitoring": "jmx",
+		"pod":        depname,
+	}
+}
+
+// getPodNames returns the pod names of the array of pods passed in
+func getPodNames(pods []corev1.Pod) []string {
+	var podNames []string
+	for _, pod := range pods {
+		podNames = append(podNames, pod.Name)
+	}
+	return podNames
+}
+
+// addNamespace adds a new NameSpace
+func (r *Wso2IsReconciler) addNamespace(m wso2v1beta1.Wso2Is) corev1.Namespace {
+	namespace := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   m.Namespace,
+			Labels: labelsForWso2IS(m.Name),
+		},
+	}
+	ctrl.SetControllerReference(&m, &namespace, r.Scheme)
+	return namespace
+}
+
+// addServiceAccount adds a new ServiceAccount
+func (r *Wso2IsReconciler) addServiceAccount(m wso2v1beta1.Wso2Is) *corev1.ServiceAccount {
+	svc := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "wso2svc-account",
+			Namespace: m.Namespace,
+		},
+	}
+	ctrl.SetControllerReference(&m, svc, r.Scheme)
+	return svc
+}
+
+// add adds a new PersistentVolume
+func (r *Wso2IsReconciler) addPersistentVolumeClaim(m wso2v1beta1.Wso2Is) *corev1.PersistentVolumeClaim {
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "user-store-pv-claim",
+			Namespace: m.Namespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				"ReadWriteMany",
+			},
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse("1Gi"),
+				},
+			},
+		},
+	}
+	ctrl.SetControllerReference(&m, pvc, r.Scheme)
+	return pvc
+}
+
+// addConfigMap adds a new ConfigMap
+func (r *Wso2IsReconciler) addConfigMap(m wso2v1beta1.Wso2Is, logger logr.Logger) *corev1.ConfigMap {
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "identity-server-conf",
+			Namespace: m.Namespace,
+		},
+		Data: map[string]string{
+			"deployment.toml": getTomlConfig(m.Spec, logger),
+		},
+	}
+	ctrl.SetControllerReference(&m, configMap, r.Scheme)
+	return configMap
+}
+
+func getTomlConfig(spec wso2v1beta1.Wso2IsSpec, logger logr.Logger) string {
+	if len(spec.TomlConfig) == 0 {
+		buf := new(bytes.Buffer)
+		if err := toml.NewEncoder(buf).Encode(spec.Configurations); err != nil {
+			log.Fatal(err)
```

Comment:
The `log.Fatal` will exit the process with error. Shall we change this to `log.Println`?

Reply:
Done